### PR TITLE
Generalise os helpers

### PR DESCRIPTION
**Pull request by rcooke-warwick**
Previously the `balenaOS` class was only usable with the image that is sent from a client. In cases where users want to download an image instead it wasn't simple to use. 

This PR allows users to provide a custom path to an OS image when creating an instance of the object.

For example, when using the OS image that is sent from the client: 
```js
this.suite.context.set({
      os: new BalenaOS(
        {
          deviceType: this.suite.deviceType.slug,
          network: this.suite.options.balenaOS.network,
          configJson: {
            uuid: this.suite.options.balenaOS.config.uuid,
            os: {
              sshKeys: [
                await this.context
                  .get()
                  .utils.createSSHKey(this.context.get().sshKeyPath),
              ],
            },
            // persistentLogging is managed by the supervisor and only read at first boot
            persistentLogging: true,
            // Set local mode so we can perform local pushes of containers to the DUT
            localMode: true,
          },
        },
        this.getLogger()
      ),
    });

 await this.context.get().os.fetch();
 await this.context.get().os.configure();
```

In the case of using an image downloaded via the sdk:
```js
this.suite.context.set({
      os: new BalenaOS(
        {
          image: `PATH_TO_IMAGE`,
          deviceType: this.suite.deviceType.slug,
          network: this.suite.options.balenaOS.network,
          configJson: {
            uuid: this.suite.options.balenaOS.config.uuid,
            os: {
              sshKeys: [
                await this.context
                  .get()
                  .utils.createSSHKey(this.context.get().sshKeyPath),
              ],
            },
            // persistentLogging is managed by the supervisor and only read at first boot
            persistentLogging: true,
            // Set local mode so we can perform local pushes of containers to the DUT
            localMode: true,
          },
        },
        this.getLogger()
      ),
    });

 await this.context.get().os.fetch();
 await this.context.get().os.configure();
```

### CODE REVIEW

**Comment by rcooke-warwick** [comment]
@vipulgupta2048 you can try using this in your HUP suite refactoring, should let you delete all of the copy pasted code from the balenaOS helper class

**Comment by vipulgupta2048** [comment]
I will try this class with Robert's test, see if it works out for the HUP test

**Comment by vipulgupta2048** [comment]
@balena-ci rebase

File: core/lib/components/os/balenaos.js [hunk 1]
```diff
@@ -84,115 +84,110 @@ const injectNetworkConfiguration = async (image, configuration) => {
...
-	unpack(download) {
-		const types = {
-			local: async () => {
-				await pipeline(
-					fs.createReadStream(download.source),
-					zlib.createGunzip(),
-					fs.createWriteStream(this.image.path),
+	// calling the fetch method will prepare the image to be used - either unzipping it or moving it to the working directory
+	async fetch(){
+		let unpack = await isGzip(this.image)
```
Comment (vipulgupta2048) [inline core/lib/components/os/balenaos.js:116]:
```suggestion
    this.logger.log('Unpacking the operating system');
		let unpack = await isGzip(this.image)
```

**Comment by vipulgupta2048** [comment]
@balena-ci rebase

**Comment by vipulgupta2048** [comment]
This is good to go.

**Comment by rcooke-warwick** [comment]
@vipulgupta2048 have you tested on both the HUP and the OS suites? If not I can check with the OS suite now

**Comment by vipulgupta2048** [comment]
The correction made by you did lead to the HUP test fail previously because it was not able to read fd correctly for some reason. I will retry HUP to see if it works now.

**Comment by vipulgupta2048** [comment]
@balena-ci rebase

**Comment by rcooke-warwick** [comment]
@vipulgupta2048 there's a couple more changes in here, worth testing it with your hup suite, and if its all good then we can merge it

**Comment by vipulgupta2048** [comment]
Sure thing, pulling and testing again

File: core/config/default.js [hunk 2]
```diff
@@ -12,6 +12,7 @@ module.exports = {
   },
   leviathan: {
     artifacts: '/tmp/artifacts',
+    downloads: '/tmp/downloads', // add an images directory in /tmp
```
Comment (vipulgupta2048) [inline core/config/default.js:15]:
```suggestion
    artifacts: '/tmp/artifacts',    // To store artifacts meant to be reported as results at the end of the suite
    downloads: '/tmp/downloads',    // To store/download assets needed for the suite (non-persistent) 
```

File: core/lib/common/suite.js [hunk 3]
```diff
@@ -248,6 +251,13 @@ class Suite {
 			path.join(config.get('leviathan.uploads.suite'), 'node_modules'),
 		);
 	}
+
+	async removeDownloads(){
+		if(fs.existsSync(config.get('leviathan.downloads'))){
```
Comment (vipulgupta2048) [inline core/lib/common/suite.js:256]:
```suggestion
		if (fs.existsSync(config.get('leviathan.downloads'))) {
```

File: core/lib/common/test.js [hunk 4]
```diff
@@ -39,8 +39,8 @@ module.exports = class Test {
 		this.id = id;
 
 		this.teardown = {
-			run: () => {
-				suite.teardown.run(this.id);
+			run: async() => {
```
Comment (vipulgupta2048) [inline core/lib/common/test.js:42]:
```suggestion
			run: async () => {
```

**Comment by rcooke-warwick** [comment]
@balena-ci I self-certify!

**Pull request merged**